### PR TITLE
Add load more button in-app

### PR DESCRIPTION
**Pull request by KokkieH**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Currently we only show a limited number of products in the in-app marketplace. We're adding the ability to load additional products to enable merchants to search/browser the full list of products available on WooCommerce.com

Closes https://github.com/Automattic/woocommerce.com/issues/21568

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Check out this branch
2. Go to `/plugins/woocommerce` and run `pnpm --filter='@woocommerce/plugin-woocommerce' build`
3. Visit the [in-app marketplace](http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions) and click on the Extensions tab
4. Scroll down to the bottom and confirm you see a Load More button
5. Click the button, and you should see additional products loaded on the page
6. If navigating via keyboard, if you press Tab after new products appear, focus should go to the first new product.
7. Confirm the same behaviour is present on the Themes tab, and also when filtering by category
8. Search for a specific term, e.g. `payments`. You should see a load more button, but it should disappear after you click it the third time (there are currently only four pages of results for that search term)
9. Confirm when switching from a paginated tab/filter/result to another option, that you're back on page 1 each time.
10. Look for anything that feels broken.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Added a Load More button to product lists on the Extensions page, to request additional search results from WooCommerce.com.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx [hunk 1]
```diff
@@ -52,7 +68,95 @@ export default function Content(): JSX.Element {
...
+						).length,
+						'business-services': combinedProducts.filter(
+							( p ) => p.type === ProductType.businessService
+						).length,
+					} );
+
+					return combinedProducts;
+				} );
+
+				speak( __( 'More products loaded', 'woocommerce' ) );
```
Comment (bor0) [inline plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:137]: 💯

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx [hunk 2]
```diff
@@ -52,7 +68,95 @@ export default function Content(): JSX.Element {
...
 	};
 
+	const loadMoreProducts = useCallback( () => {
+		setIsLoadingMore( true );
+		const params = new URLSearchParams();
+		const abortController = new AbortController();
+
+		if ( query.category && query.category !== '_all' ) {
+			params.append( 'category', query.category );
+		}
```
Comment (bor0) [inline plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:79]: For some reason `query.category` is undefined when I'm on the themes tab, and instead of searching for themes it searches for extensions

https://github.com/user-attachments/assets/6faeced0-1ab9-4d22-a872-9bc0a1eb7c3c

Comment (bor0) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:79]: Possibly related issue: the product state is maintained when I switch tabs.

Scenario:
1. Open page
2. Click load more in themes (page -> 2)
3. Switch to extensions tab, click load more in extensions (page -> 3)

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx [hunk 3]
```diff
@@ -52,7 +68,95 @@ export default function Content(): JSX.Element {
...
+		query.term,
+		setIsLoadingMore,
+		setSearchResultsCount,
+	] );
+
 	useEffect( () => {
+		// if it's a paginated request, don't use this effect
+		if ( currentPage > 1 ) {
+			return;
+		}
```
Comment (bor0) [inline plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:164]: I see the value of `currentPage` is only affected within `loadMoreProducts` - we may want to reset it when switching tabs? And when introducing a search term? Any other scenarios?

Comment (bor0) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:164]: Scratch this, I see we're doing it [here](https://github.com/woocommerce/woocommerce/pull/51434/files#diff-66d84dd7feb9389b0808515aed92c9eebc6bad572bd192d6ca327bee3b6796d1R343).

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx [hunk 4]
```diff
@@ -212,30 +340,46 @@ export default function Content(): JSX.Element {
...
+			case 'themes':
+				return ProductType.theme;
+			case 'business-services':
+				return ProductType.businessService;
+			default:
+				return ProductType.extension;
+		}
+	};
+
+	useEffect( () => {
```
Comment (bor0) [inline plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:356]: This is beautiful.

```suggestion
	// Maintain product focus for accessibility.
	useEffect( () => {
```

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 2 above)
Comment (KokkieH) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:79]: Ah, I see what's happening. I'm checking for `query.category`, but when switching between the tabs `query.tab` is changed, and a category query param is only added if you click on a category filter.

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 3 above)
Comment (KokkieH) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:164]: This is taken care of by the useEffect hook on line 344

File: plugins/woocommerce-admin/client/marketplace/components/load-more-button/load-more-button.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+	return (
+		<Button
+			className="woocommerce-marketplace__load-more"
+			variant={ 'secondary' }
+			onClick={ handleClick }
+			isBusy={ isBusy }
+			disabled={ disabled }
+		>
+			{ isBusy && speak( __( 'Loading more products', 'woocommerce' ) ) }
```
Comment (bor0) [inline plugins/woocommerce-admin/client/marketplace/components/load-more-button/load-more-button.tsx]: Does it need to go to the JSX part?

```suggestion
	}

	if ( isBusy ) {
		speak( __( 'Loading more products', 'woocommerce' ) );
	}

	return (
		<Button
			className="woocommerce-marketplace__load-more"
			variant={ 'secondary' }
			onClick={ handleClick }
			isBusy={ isBusy }
			disabled={ disabled }
		>
```

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 2 above)
Comment (KokkieH) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:79]: Should be fixed by https://github.com/woocommerce/woocommerce/pull/51434/commits/a26a03d7e68fc32525e2e1029e14d5319d58e8b7

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 4 above)
Comment (KokkieH) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:356]: added in https://github.com/woocommerce/woocommerce/pull/51434/commits/a26a03d7e68fc32525e2e1029e14d5319d58e8b7

File: plugins/woocommerce-admin/client/marketplace/components/load-more-button/load-more-button.tsx (hunk 5 above)
Comment (KokkieH) [reply plugins/woocommerce-admin/client/marketplace/components/load-more-button/load-more-button.tsx]: It doesn't, no. Good suggestion.

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx [hunk 6]
```diff
@@ -212,30 +340,46 @@ export default function Content(): JSX.Element {
...
+	const getProductType = ( tab: string ): ProductType => {
+		switch ( tab ) {
+			case 'themes':
+				return ProductType.theme;
+			case 'business-services':
+				return ProductType.businessService;
+			default:
+				return ProductType.extension;
+		}
+	};
```
Comment (bor0) [inline plugins/woocommerce-admin/client/marketplace/components/content/content.tsx]: Should this go into `utils/functions.tsx`? Doesn't necessarily need to be done as part of this PR but I feel like we can improve the whole `ProductType` and string conversion flow

Comment (bor0) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx]: On related note, [this](https://github.com/woocommerce/woocommerce/blob/554434ea3d366ce688c3d945c2e617cf83a32554/plugins/woocommerce-admin/client/marketplace/components/category-selector/category-selector.tsx#L20) feels weird. However, it'd require an API change as well, since these work:

- https://woocommerce.com/wp-json/wccom-extensions/1.0/search?country=MK (extensions)
- https://woocommerce.com/wp-json/wccom-extensions/1.0/search?category=themes&country=MK (themes)

But this doesn't:
- https://woocommerce.com/wp-json/wccom-extensions/1.0/search?category=extensions&country=MK (extensions)

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 2 above)
Comment (bor0) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:79]: Works. Thanks!

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx [hunk 7]
```diff
@@ -266,6 +435,13 @@ export default function Content(): JSX.Element {
 			) }
 
 			{ renderContent() }
+			{ shouldShowLoadMoreButton() && (
```
Comment (bor0) [inline plugins/woocommerce-admin/client/marketplace/components/content/content.tsx]: Perhaps,

```suggestion
			{ ! isLoading && shouldShowLoadMoreButton() && (
```

Or `if ( isLoading ) return false` within `shouldShowLoadMoreButton`.

(If you accept the suggestion, don't forget to extract `isLoading` from `marketplaceContextValue`)

I'm seeing this edge-case (you have to be super fast to reproduce it :))


https://github.com/user-attachments/assets/5fdcc3db-ca4f-45a5-8cff-1467e9a2f4b3

File: plugins/woocommerce-admin/client/marketplace/components/load-more-button/load-more-button.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+	return (
+		<Button
+			className="woocommerce-marketplace__load-more"
+			variant={ 'secondary' }
+			onClick={ handleClick }
+			isBusy={ isBusy }
+			disabled={ disabled }
+		>
+			{ __( 'Load more', 'woocommerce' ) }
```
Comment (bor0) [inline plugins/woocommerce-admin/client/marketplace/components/load-more-button/load-more-button.tsx:34]: This is fine for now, athough we chatted about infinite loading on the code chat call today.

Maybe we can create an issue about it to see if it's worth pursuing? May improve UX and especially discoverability (h/t @andfinally)

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 6 above)
Comment (bor0) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx]: Can create an issue about it and unblock this PR.

**Review by bor0 — APPROVED** [review]
Left a small edge-case note, but in general this works and tests well! Left a few more suggestions which can be done as part of different issues to unblock this.

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 7 above)
Comment (KokkieH) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx]: Ha! I had that check, and removed it after the changes we made in the code chat. Added back in aadaeb3

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 6 above)
Comment (KokkieH) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx]: > Should this go into utils/functions.tsx? Doesn't necessarily need to be done as part of this PR but I feel like we can improve the whole ProductType and string conversion flow

It could, yeah. I had actually looked if I couldn't replace some of the other ProductType checks in the component with this function, but it doesn't look that way. Moved the function in aadaeb3

> On related note, [this](https://github.com/woocommerce/woocommerce/blob/554434ea3d366ce688c3d945c2e617cf83a32554/plugins/woocommerce-admin/client/marketplace/components/category-selector/category-selector.tsx#L20) feels weird. However, it'd require an API change as well, since these work:

Yeah, I suspect originally this was designed to only support extensions, and then themes were added afterwards. To fix on the API side will likely mean we need to support two versions of the endpoint for backwards compatibility.

**Comment by bor0** [comment]
@KokkieH when you merge this into the feature branch, don't forget to add it as an item-to-test within #51413. Let's see if we can squeeze that PR in before the next code freeze :)

🚢🚢🚢🚢
